### DISCUSSION TITLE
Counting Steps for Patching.MaxStepsForScript

### DISCUSSION THREAD

**Discussion by mic2zar** [ravendb/ravendb#17715] [Q&A] [answered]
Could you please clarify how are you counting Patching.MaxStepsForScript. We tried a couple of simple experiments with this parameter (default is 10K), but our findings were not that conclusive. The simplest example, which in our opinion should **not** work if we have more documents than 10K is the following:
```
from SomeCollection as a
{
    a.Test = 5;   
}
```
but still works. At the same time, once we put some iterations for collections inside of this code, we reach finally the level that we fail. How could we figure out what would be counted in the script as _Step_ and what will be not?

**Comment by garayx** [comment]
Hi

RavenDB uses the [Jint](https://github.com/sebastienros/jint) engine under the hood for patching, 

The `Patching.MaxStepsForScript` limit is passed to the `MaxStatements` configuration option of the Jint engine:
https://github.com/sebastienros/jint#execution-constraints

I am not sure how `MaxStatements` limit is calculated inside Jint,

**Comment by lahma** [comment]
I belive that Jint doesn't know about the `from SomeCollection as a` part, so RavenDB will feed all the documents in that collection into projection and then Jint kicks in. Jin will only observe "here's a document, project it as follows using JS engine".

So in this case things happening inside the curly braces are counted towards statements, like the assignment in the example.

**Reply by mic2zar** [reply]
By then as per my example from the first message with 20K documents in SomeCollection I would have 20K assignments of value 5 to property Test. Still for that case the default constraint of 10K for MaxStepsForScript does not fire.

**Reply by grisha-kotler** [reply, marked as answer]
The limitation is per document.
In order to trigger it, you need something like:
```
from "Orders" as o update {
    for (var i = 0; i < 10000; i++)
        o.Test++;
}
```
